// @enhavo/user/manager/UserManager.ts
import {Router} from "@enhavo/app/routing/Router";
import {Form} from "@enhavo/vue-form/model/Form";
import {FormFactory} from "@enhavo/vue-form/form/FormFactory";
import {FormUtil} from "@enhavo/vue-form/form/FormUtil";

export default class UserManager
{
    public loginForm: Form
    public loading: boolean;

    constructor(
        public branding: Branding,
        private router: Router,
        private formFactory: FormFactory,
    ) {

    }

    public loadLogin()
    {
        this.loading = true;
        const url = this.router.generate('enhavo_user_api_login');
        fetch(url).then((response) => {
            response.json().then((data: any) => {
                this.loading = false;
                this.loginForm = this.formFactory.create(data.form);
                console.log(this.loginForm);
            });
        });
    }

    public login()
    {
        this.loading = true;
        const url = this.router.generate('enhavo_user_api_login');
        const data = FormUtil.serializeForm(this.loginForm);
        console.log(data)
        fetch(url, {
            method: 'POST',
            body: data,
        }).then((response) => {
            response.json().then((data: any) => {
                if (data.success) {
                    // @ts-ignore
                    window.location.href = data.redirect;
                    return;
                }

                this.loading = false;
                this.loginForm = this.formFactory.create(data.form);
            });
        });
    }
}

class Branding
{
    enableCreatedBy: boolean;
    enableVersion: boolean;
    enable: boolean;
    version: string;
    text: string;
    logo: string;
    backgroundImage: string;
}
