// @enhavo/media-library/MediaLibrary.ts
import axios, {CancelTokenSource} from "axios";
import * as _ from "lodash";
import Data, {Column, ContentType, File, Tag} from "@enhavo/media-library/Data";
import UpdatedEvent from "@enhavo/app/view-stack/event/UpdatedEvent";
import * as async from "async";
import Message from "@enhavo/app/flash-message/Message";
import FlashMessenger from "@enhavo/app/flash-message/FlashMessenger";
import EventDispatcher from "@enhavo/app/view-stack/EventDispatcher";
import View from "@enhavo/app/view/View";
import Router from "@enhavo/core/Router";
import ComponentRegistryInterface from "@enhavo/core/ComponentRegistryInterface";
import ViewInterface from "@enhavo/app/view-stack/ViewInterface";
import Translator from "@enhavo/core/Translator";
import CloseEvent from "@enhavo/app/view-stack/event/CloseEvent";

export default class MediaLibrary {
    public data: Data;
    private source: CancelTokenSource;
    private readonly eventDispatcher: EventDispatcher;
    private readonly flashMessenger: FlashMessenger;
    private readonly view: View;
    private readonly router: Router;
    private readonly componentRegistry: ComponentRegistryInterface;
    private readonly translator: Translator;

    public constructor(data: any, eventDispatcher: EventDispatcher, view: View, flashMessenger: FlashMessenger, router: Router, componentRegistry: ComponentRegistryInterface, translator: Translator) {
        this.data = _.extend(data.data, new Data());
        this.flashMessenger = flashMessenger;
        this.eventDispatcher = eventDispatcher;
        this.view = view;
        this.router = router;
        this.componentRegistry = componentRegistry;
        this.translator = translator;

        $(document).on('add', () => {
            this.add();
        });
    }

    public init() {
        this.componentRegistry.registerStore('mediaLibrary', this);
        this.data = this.componentRegistry.registerData(this.data);

        this.getTags();
        this.getContentTypes();
        this.getFiles();
    }

    private add() {
        this.view.loading();
        let calls: ((callback: (err: any) => void) => void)[] = [];
        for (let image of this.data.selectedFiles) {
            calls.push((callback) => {
                axios
                    .get(this.router.generate('enhavo_media_library_file_add', {id: image.id}))
                    .then(response => {
                        this.eventDispatcher.dispatch(new UpdatedEvent(this.view.getId(), response.data));
                        callback(null);
                    })
                    .catch(error => {
                        callback('error');
                    })
            });
        }

        async.parallel(calls, (err) => {
            if (err) {
                this.flashMessenger.addMessage(new Message(Message.ERROR, 'media.library.add.error'));
            } else {
                this.flashMessenger.addMessage(new Message(Message.SUCCESS, 'media.library.add.success'));
                this.eventDispatcher.dispatch(new CloseEvent(this.view.getId())); // todo: could be a setting
            }
            this.view.loaded();
            this.data.selectedFiles = [];
        });
    }

    public loading() {
        this.data.loading = true;
    }

    public loaded() {
        this.data.loading = false;
    }

    public setProgress(value: number) {
        this.data.progress = value;
    }

    public open(file: File): void {
        if (this.data.mode === 'edit') {
            this.activateItem(file).then(() => {
                let url = this.router.generate('enhavo_media_library_file_update', {
                    id: file.id,
                });
                this.view.open(url, 'edit-view').then((view: ViewInterface) => {
                    this.view.storeValue('active-view', view.id);
                });
            });

        } else if (this.data.mode === 'select') {
            if (!this.isFileSelected(file)) {
                if (this.data.multiple) {
                    this.addItem(file);
                } else {
                    this.data.selectedFiles = [];
                    this.addItem(file);
                }
            } else {
                this.removeItem(file);
            }

            this.checkFileSelection();
        }
    }

    private activateItem(item: File) {
        return new Promise((resolve, reject) => {
            async.parallel([(callback: (err: any) => void) => {
                this.view.storeValue('active-view', null).then(() => {
                    callback(null);
                }).catch(() => {
                    callback(true);
                });
            }], (err) => {
                if (err) {
                    reject();
                } else {
                    resolve();
                }
            });
        });
    }

    private checkFileSelection() {
        for (let file of this.data.files) {
            file.selected = this.isFileSelected(file);
        }
    }

    public isFileSelected(item: File): boolean {
        for (let selectedItem of this.data.selectedFiles) {
            if (selectedItem.id == item.id) {
                return true;
            }
        }
        return false;
    }

    private removeItem(item: File) {
        let index: number = -1;
        for (let i in this.data.selectedFiles) {
            if (this.data.selectedFiles[i].id == item.id) {
                index = parseInt(i);
                break;
            }
        }

        if (index !== -1) {
            this.data.selectedFiles.splice(index, 1);
        }
    }

    private addItem(item: any) {
        this.data.selectedFiles.push(item);
    }

    public setActiveTag(tag: Tag) {
        if (this.data.activeTag) {
            this.data.activeTag.selected = false;
        }
        if (this.data.activeTag === tag) {
            this.data.activeTag.selected = false;
            this.data.activeTag = null;
        } else {
            this.data.activeTag = tag;
            this.data.activeTag.selected = true;
        }

        this.setActivePage(1);
    }

    public setActiveContentType(contentType: ContentType) {
        if (this.data.activeContentType) {
            this.data.activeContentType.selected = false;
        }
        if (this.data.activeContentType === contentType) {
            this.data.activeContentType.selected = false;
            this.data.activeContentType = null;
        } else {
            this.data.activeContentType = contentType;
            this.data.activeContentType.selected = true;
        }

        this.setActivePage(1);
    }

    public setSortColumn(column: Column) {
        if (this.data.sortColumn && this.data.sortColumn.property == column.property) {
            column.direction = column.direction == 'asc' ? 'desc' : 'asc';
        } else {
            column.direction = 'asc';
        }
        this.data.sortColumn = column;

        this.getFiles();
    }

    public fail(message: string) {
        this.flashMessenger.addMessage(new Message(Message.ERROR, message));
    }

    public refresh() {
        // todo: load tags and reselect?
        this.getFiles();
    }

    public setActivePage(page: number) {
        this.data.page = page;

        this.getFiles();
    }

    private getFiles() {
        this.loading();
        let url = this.router.generate('enhavo_media_library_file_files', {
            page: this.data.page,
            limit: this.data.limit,
        });

        if (this.source != null) {
            this.source.cancel();
        }
        this.source = axios.CancelToken.source();
        axios
            .post(url, {
                filters: [
                    {name: 'contentType', value: this.data.activeContentType ? this.data.activeContentType.key : null},
                    {name: 'tags', value: this.data.activeTag ? [this.data.activeTag.id] : null},
                    {name: 'filename', value: this.data.searchString}
                ],
                sorting: this.data.sortColumn ? [
                    {property: this.data.sortColumn.property, direction: this.data.sortColumn.direction}
                ] : null
            }, {
                cancelToken: this.source.token
            })
            .then(response => {
                this.source = null;

                this.data.files = [];
                let filesData = response.data.files;
                for (let fileData of filesData) {
                    let file = new File();
                    file.id = fileData.id;
                    file.label = fileData.label;
                    file.suffix = fileData.suffix;
                    file.date = fileData.date;
                    file.type = fileData.type;
                    file.icon = fileData.icon;
                    file.previewImageUrl = fileData.previewImageUrl;
                    this.data.files.push(file);
                }

                this.data.pages = response.data.pages;
                this.data.columns = [];
                for (let columnData of response.data.columns) {
                    let column = new Column();
                    column.property = columnData.property;
                    column.label = columnData.label;
                    this.data.columns.push(column);
                }
                this.data.page = response.data.page;

                if (!this.data.sortColumn) {
                    this.data.sortColumn = this.data.columns[0];
                }
                this.checkFileSelection();
                this.loaded();
            });
    }

    private getTags() {
        this.loading();
        let url = this.router.generate('enhavo_media_library_file_tags');

        axios
            .get(url, {
                params: null,
            })
            .then(response => {
                let tags = [];
                let tagsData = response.data.tags;
                for (let tagData of tagsData) {
                    let tag = new Tag();
                    tag.id = tagData.id;
                    tag.label = tagData.label;
                    tags.push(tag);
                }
                this.data.tags = tags;
                this.loaded();
            });
    }

    private getContentTypes() {
        this.loading();
        let url = this.router.generate('enhavo_media_library_file_content_types');

        axios
            .get(url, {
                params: null,
            })
            .then(response => {
                let contentTypes = [];
                let contentTypesData = response.data.content_types;
                for (let contentTypeData of contentTypesData) {
                    let contentType = new ContentType();
                    contentType.key = contentTypeData.key;
                    contentType.label = contentTypeData.label;
                    contentTypes.push(contentType);
                }
                this.data.contentTypes = contentTypes;
                this.loaded();
            });
    }

    public search() {
        this.setActivePage(1);
    }

    public clearSearch() {
        this.data.searchString = '';
        this.setActivePage(1);
    }


    public showDropZone() {
        if (!this.data.dropZone) {
            this.data.dropZone = true;
        }
    }

    public showDropZoneActive() {
        if (!this.data.dropZoneActive) {
            this.data.dropZoneActive = true;
        }
    }

    public hideDropZone() {
        if (this.data.dropZone) {
            this.data.dropZone = false;
        }
    }

    public hideDropZoneActive() {
        if (this.data.dropZoneActive) {
            this.data.dropZoneActive = false;
        }
    }

    public getSortDirection() {
        return this.data.sortColumn ? this.data.sortColumn.direction : 'asc';
    }

    public setView(type: string) {
        this.data.view = type;
    }
}
