// @enhavo/form/wysiwyg/WysiwygSpecialCharactersButton.ts
import {WysiwygMenuButton} from "@enhavo/form/wysiwyg/WysiwygMenuButton";
import {WysiwygForm} from "@enhavo/form/form/model/WysiwygForm";
import {Translator} from "@enhavo/app/translation/Translator";

export class WysiwygSpecialCharactersButton extends WysiwygMenuButton
{
    component: string = 'form-wysiwyg-menu-button';

    click = (event: Event, form: WysiwygForm) => {
        form.openModal('form-wysiwyg-modal-special-characters', {
            configuration: this,
        })
            .then((result: object) => {
                if (result.hasOwnProperty('character')) {
                    form.editor.chain().focus().insertContent(result.character).run();
                }
            })
            .catch(() => {
                // Cancelled
            })
        ;
    };

    characters: CharacterTab[] = [
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.dollar_sign',
                    searchTags: '',
                    code: 36,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.cent_sign',
                    searchTags: '',
                    code: 162,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.euro_sign',
                    searchTags: '',
                    code: 8364,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.pound_sign',
                    searchTags: '',
                    code: 163,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.yen_sign',
                    searchTags: '',
                    code: 165,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.currency_sign',
                    searchTags: '',
                    code: 164,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.euro_currency_sign',
                    searchTags: '',
                    code: 8352,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.colon_sign',
                    searchTags: '',
                    code: 8353,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.cruzeiro_sign',
                    searchTags: '',
                    code: 8354,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.french_franc_sign',
                    searchTags: '',
                    code: 8355,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.lira_sign',
                    searchTags: '',
                    code: 8356,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.mill_sign',
                    searchTags: '',
                    code: 8357,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.naira_sign',
                    searchTags: '',
                    code: 8358,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.peseta_sign',
                    searchTags: '',
                    code: 8359,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.rupee_sign',
                    searchTags: '',
                    code: 8360,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.won_sign',
                    searchTags: '',
                    code: 8361,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.new_sheqel_sign',
                    searchTags: '',
                    code: 8362,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.dong_sign',
                    searchTags: '',
                    code: 8363,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.kip_sign',
                    searchTags: '',
                    code: 8365,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.tugrik_sign',
                    searchTags: '',
                    code: 8366,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.drachma_sign',
                    searchTags: '',
                    code: 8367,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.german_penny_symbol',
                    searchTags: '',
                    code: 8368,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.peso_sign',
                    searchTags: '',
                    code: 8369,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.guarani_sign',
                    searchTags: '',
                    code: 8370,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.austral_sign',
                    searchTags: '',
                    code: 8371,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.hryvnia_sign',
                    searchTags: '',
                    code: 8372,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.cedi_sign',
                    searchTags: '',
                    code: 8373,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.livre_tournois_sign',
                    searchTags: '',
                    code: 8374,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.spesmilo_sign',
                    searchTags: '',
                    code: 8375,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.tenge_sign',
                    searchTags: '',
                    code: 8376,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.indian_rupee_sign',
                    searchTags: '',
                    code: 8377,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.turkish_lira_sign',
                    searchTags: '',
                    code: 8378,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.nordic_mark_sign',
                    searchTags: '',
                    code: 8379,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.manat_sign',
                    searchTags: '',
                    code: 8380,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.ruble_sign',
                    searchTags: '',
                    code: 8381,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.yen_character',
                    searchTags: '',
                    code: 20870,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.yuan_character',
                    searchTags: '',
                    code: 20803,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.yuan_character,_in_hong_kong_and_taiwan',
                    searchTags: '',
                    code: 22291,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ccurrency.yen/yuan_character_variant_one',
                    searchTags: '',
                    code: 22278,
                    duplicate: false,
                },
            ],
        },
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.copyright_sign',
                    searchTags: '',
                    code: 169,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.registered_sign',
                    searchTags: '',
                    code: 174,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.trade_mark_sign',
                    searchTags: '',
                    code: 8482,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.per_mille_sign',
                    searchTags: '',
                    code: 8240,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.micro_sign',
                    searchTags: '',
                    code: 181,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.middle_dot',
                    searchTags: '',
                    code: 183,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.bullet',
                    searchTags: '',
                    code: 8226,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.three_dot_leader',
                    searchTags: '',
                    code: 8230,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.minutes_/_feet',
                    searchTags: '',
                    code: 8242,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.seconds_/_inches',
                    searchTags: '',
                    code: 8243,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.section_sign',
                    searchTags: '',
                    code: 167,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.paragraph_sign',
                    searchTags: '',
                    code: 182,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ttext.sharp_s_/_ess_zed',
                    searchTags: '',
                    code: 223,
                    duplicate: false,
                },
            ],
        },
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.single_left_pointing_angle_quotation_mark',
                    searchTags: '',
                    code: 8249,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.single_right_pointing_angle_quotation_mark',
                    searchTags: '',
                    code: 8250,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.left_pointing_guillemet',
                    searchTags: '',
                    code: 171,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.right_pointing_guillemet',
                    searchTags: '',
                    code: 187,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.left_single_quotation_mark',
                    searchTags: '',
                    code: 8216,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.right_single_quotation_mark',
                    searchTags: '',
                    code: 8217,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.left_double_quotation_mark',
                    searchTags: '',
                    code: 8220,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.right_double_quotation_mark',
                    searchTags: '',
                    code: 8221,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.single_low_9_quotation_mark',
                    searchTags: '',
                    code: 8218,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.double_low_9_quotation_mark',
                    searchTags: '',
                    code: 8222,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.less_than_sign',
                    searchTags: '',
                    code: 60,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.greater_than_sign',
                    searchTags: '',
                    code: 62,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.less_than_or_equal_to',
                    searchTags: '',
                    code: 8804,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.greater_than_or_equal_to',
                    searchTags: '',
                    code: 8805,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.en_dash',
                    searchTags: '',
                    code: 8211,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.em_dash',
                    searchTags: '',
                    code: 8212,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.macron',
                    searchTags: '',
                    code: 175,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.overline',
                    searchTags: '',
                    code: 8254,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.currency_sign',
                    searchTags: '',
                    code: 164,
                    duplicate: true,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.broken_bar',
                    searchTags: '',
                    code: 166,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.diaeresis',
                    searchTags: '',
                    code: 168,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.inverted_exclamation_mark',
                    searchTags: '',
                    code: 161,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.turned_question_mark',
                    searchTags: '',
                    code: 191,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.circumflex_accent',
                    searchTags: '',
                    code: 710,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.small_tilde',
                    searchTags: '',
                    code: 732,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.degree_sign',
                    searchTags: '',
                    code: 176,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.minus_sign',
                    searchTags: '',
                    code: 8722,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.plus_minus_sign',
                    searchTags: '',
                    code: 177,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.division_sign',
                    searchTags: '',
                    code: 247,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.fraction_slash',
                    searchTags: '',
                    code: 8260,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.multiplication_sign',
                    searchTags: '',
                    code: 215,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.superscript_one',
                    searchTags: '',
                    code: 185,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.superscript_two',
                    searchTags: '',
                    code: 178,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.superscript_three',
                    searchTags: '',
                    code: 179,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.fraction_one_quarter',
                    searchTags: '',
                    code: 188,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.fraction_one_half',
                    searchTags: '',
                    code: 189,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.qquotations.fraction_three_quarters',
                    searchTags: '',
                    code: 190,
                    duplicate: false,
                },
            ],
        },
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.function_/_florin',
                    searchTags: '',
                    code: 402,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.integral',
                    searchTags: '',
                    code: 8747,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.n_ary_sumation',
                    searchTags: '',
                    code: 8721,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.infinity',
                    searchTags: '',
                    code: 8734,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.square_root',
                    searchTags: '',
                    code: 8730,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.similar_to',
                    searchTags: '',
                    code: 8764,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.approximately_equal_to',
                    searchTags: '',
                    code: 8773,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.almost_equal_to',
                    searchTags: '',
                    code: 8776,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.not_equal_to',
                    searchTags: '',
                    code: 8800,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.identical_to',
                    searchTags: '',
                    code: 8801,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.element_of',
                    searchTags: '',
                    code: 8712,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.not_an_element_of',
                    searchTags: '',
                    code: 8713,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.contains_as_member',
                    searchTags: '',
                    code: 8715,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.n_ary_product',
                    searchTags: '',
                    code: 8719,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.logical_and',
                    searchTags: '',
                    code: 8743,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.logical_or',
                    searchTags: '',
                    code: 8744,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.not_sign',
                    searchTags: '',
                    code: 172,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.intersection',
                    searchTags: '',
                    code: 8745,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.union',
                    searchTags: '',
                    code: 8746,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.partial_differential',
                    searchTags: '',
                    code: 8706,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.for_all',
                    searchTags: '',
                    code: 8704,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.there_exists',
                    searchTags: '',
                    code: 8707,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.diameter',
                    searchTags: '',
                    code: 8709,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.backward_difference',
                    searchTags: '',
                    code: 8711,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.asterisk_operator',
                    searchTags: '',
                    code: 8727,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.proportional_to',
                    searchTags: '',
                    code: 8733,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.mmathematical.angle',
                    searchTags: '',
                    code: 8736,
                    duplicate: false,
                },
            ],
        },
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_grave',
                    searchTags: '',
                    code: 192,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_acute',
                    searchTags: '',
                    code: 193,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_circumflex',
                    searchTags: '',
                    code: 194,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_tilde',
                    searchTags: '',
                    code: 195,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_diaeresis',
                    searchTags: '',
                    code: 196,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_ring_above',
                    searchTags: '',
                    code: 197,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aa_macron',
                    searchTags: '',
                    code: 256,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ligature_aaee',
                    searchTags: '',
                    code: 198,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.cc_cedilla',
                    searchTags: '',
                    code: 199,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ee_grave',
                    searchTags: '',
                    code: 200,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ee_acute',
                    searchTags: '',
                    code: 201,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ee_circumflex',
                    searchTags: '',
                    code: 202,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ee_diaeresis',
                    searchTags: '',
                    code: 203,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ee_macron',
                    searchTags: '',
                    code: 274,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ii_grave',
                    searchTags: '',
                    code: 204,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ii_acute',
                    searchTags: '',
                    code: 205,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ii_circumflex',
                    searchTags: '',
                    code: 206,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ii_diaeresis',
                    searchTags: '',
                    code: 207,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ii_macron',
                    searchTags: '',
                    code: 298,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.eetthh',
                    searchTags: '',
                    code: 208,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.nn_tilde',
                    searchTags: '',
                    code: 209,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_grave',
                    searchTags: '',
                    code: 210,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_acute',
                    searchTags: '',
                    code: 211,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_circumflex',
                    searchTags: '',
                    code: 212,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_tilde',
                    searchTags: '',
                    code: 213,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_diaeresis',
                    searchTags: '',
                    code: 214,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_slash',
                    searchTags: '',
                    code: 216,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oo_macron',
                    searchTags: '',
                    code: 332,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ligature_ooee',
                    searchTags: '',
                    code: 338,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ss_caron',
                    searchTags: '',
                    code: 352,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.uu_grave',
                    searchTags: '',
                    code: 217,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.uu_acute',
                    searchTags: '',
                    code: 218,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.uu_circumflex',
                    searchTags: '',
                    code: 219,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.uu_diaeresis',
                    searchTags: '',
                    code: 220,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.uu_macron',
                    searchTags: '',
                    code: 362,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.yy_acute',
                    searchTags: '',
                    code: 221,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.yy_diaeresis',
                    searchTags: '',
                    code: 376,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.yy_macron',
                    searchTags: '',
                    code: 562,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.tthhoorrnn',
                    searchTags: '',
                    code: 222,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_grave',
                    searchTags: '',
                    code: 224,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_acute',
                    searchTags: '',
                    code: 225,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_circumflex',
                    searchTags: '',
                    code: 226,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_tilde',
                    searchTags: '',
                    code: 227,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_diaeresis',
                    searchTags: '',
                    code: 228,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_ring_above',
                    searchTags: '',
                    code: 229,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.a_macron',
                    searchTags: '',
                    code: 257,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ligature_ae',
                    searchTags: '',
                    code: 230,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.c_cedilla',
                    searchTags: '',
                    code: 231,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.e_grave',
                    searchTags: '',
                    code: 232,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.e_acute',
                    searchTags: '',
                    code: 233,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.e_circumflex',
                    searchTags: '',
                    code: 234,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.e_diaeresis',
                    searchTags: '',
                    code: 235,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.e_macron',
                    searchTags: '',
                    code: 275,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.i_grave',
                    searchTags: '',
                    code: 236,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.i_acute',
                    searchTags: '',
                    code: 237,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.i_circumflex',
                    searchTags: '',
                    code: 238,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.i_diaeresis',
                    searchTags: '',
                    code: 239,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.i_macron',
                    searchTags: '',
                    code: 299,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.eth',
                    searchTags: '',
                    code: 240,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.n_tilde',
                    searchTags: '',
                    code: 241,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_grave',
                    searchTags: '',
                    code: 242,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_acute',
                    searchTags: '',
                    code: 243,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_circumflex',
                    searchTags: '',
                    code: 244,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_tilde',
                    searchTags: '',
                    code: 245,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_diaeresis',
                    searchTags: '',
                    code: 246,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_slash',
                    searchTags: '',
                    code: 248,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.o_macron',
                    searchTags: '',
                    code: 333,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ligature_oe',
                    searchTags: '',
                    code: 339,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.s_caron',
                    searchTags: '',
                    code: 353,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.u_grave',
                    searchTags: '',
                    code: 249,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.u_acute',
                    searchTags: '',
                    code: 250,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.u_circumflex',
                    searchTags: '',
                    code: 251,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.u_diaeresis',
                    searchTags: '',
                    code: 252,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.u_macron',
                    searchTags: '',
                    code: 363,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.y_acute',
                    searchTags: '',
                    code: 253,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.thorn',
                    searchTags: '',
                    code: 254,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.y_diaeresis',
                    searchTags: '',
                    code: 255,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.y_macron',
                    searchTags: '',
                    code: 563,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.aalpha',
                    searchTags: '',
                    code: 913,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.bbeta',
                    searchTags: '',
                    code: 914,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ggamma',
                    searchTags: '',
                    code: 915,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ddelta',
                    searchTags: '',
                    code: 916,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.eepsilon',
                    searchTags: '',
                    code: 917,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.zzeta',
                    searchTags: '',
                    code: 918,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.eeta',
                    searchTags: '',
                    code: 919,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ttheta',
                    searchTags: '',
                    code: 920,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.iiota',
                    searchTags: '',
                    code: 921,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.kkappa',
                    searchTags: '',
                    code: 922,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.llambda',
                    searchTags: '',
                    code: 923,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.mmu',
                    searchTags: '',
                    code: 924,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.nnu',
                    searchTags: '',
                    code: 925,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.xxi',
                    searchTags: '',
                    code: 926,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oomicron',
                    searchTags: '',
                    code: 927,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ppi',
                    searchTags: '',
                    code: 928,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.rrho',
                    searchTags: '',
                    code: 929,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ssigma',
                    searchTags: '',
                    code: 931,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ttau',
                    searchTags: '',
                    code: 932,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.uupsilon',
                    searchTags: '',
                    code: 933,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.pphi',
                    searchTags: '',
                    code: 934,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.cchi',
                    searchTags: '',
                    code: 935,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.ppsi',
                    searchTags: '',
                    code: 936,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.oomega',
                    searchTags: '',
                    code: 937,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.alpha',
                    searchTags: '',
                    code: 945,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.beta',
                    searchTags: '',
                    code: 946,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.gamma',
                    searchTags: '',
                    code: 947,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.delta',
                    searchTags: '',
                    code: 948,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.epsilon',
                    searchTags: '',
                    code: 949,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.zeta',
                    searchTags: '',
                    code: 950,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.eta',
                    searchTags: '',
                    code: 951,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.theta',
                    searchTags: '',
                    code: 952,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.iota',
                    searchTags: '',
                    code: 953,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.kappa',
                    searchTags: '',
                    code: 954,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.lambda',
                    searchTags: '',
                    code: 955,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.mu',
                    searchTags: '',
                    code: 956,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.nu',
                    searchTags: '',
                    code: 957,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.xi',
                    searchTags: '',
                    code: 958,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.omicron',
                    searchTags: '',
                    code: 959,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.pi',
                    searchTags: '',
                    code: 960,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.rho',
                    searchTags: '',
                    code: 961,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.final_sigma',
                    searchTags: '',
                    code: 962,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.sigma',
                    searchTags: '',
                    code: 963,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.tau',
                    searchTags: '',
                    code: 964,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.upsilon',
                    searchTags: '',
                    code: 965,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.phi',
                    searchTags: '',
                    code: 966,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.chi',
                    searchTags: '',
                    code: 967,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.psi',
                    searchTags: '',
                    code: 968,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.eextended_llatin.omega',
                    searchTags: '',
                    code: 969,
                    duplicate: false,
                },
            ],
        },
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.alef_symbol',
                    searchTags: '',
                    code: 8501,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.pi_symbol',
                    searchTags: '',
                    code: 982,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.real_part_symbol',
                    searchTags: '',
                    code: 8476,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.upsilon_hook_symbol',
                    searchTags: '',
                    code: 978,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.wweierstrass_p',
                    searchTags: '',
                    code: 8472,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.ssymbols.imaginary_part',
                    searchTags: '',
                    code: 8465,
                    duplicate: false,
                },
            ],
        },
        {
            label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.label',
            characters: [
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.leftwards_arrow',
                    searchTags: '',
                    code: 8592,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.upwards_arrow',
                    searchTags: '',
                    code: 8593,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.rightwards_arrow',
                    searchTags: '',
                    code: 8594,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.downwards_arrow',
                    searchTags: '',
                    code: 8595,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.left_right_arrow',
                    searchTags: '',
                    code: 8596,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.carriage_return',
                    searchTags: '',
                    code: 8629,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.leftwards_double_arrow',
                    searchTags: '',
                    code: 8656,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.upwards_double_arrow',
                    searchTags: '',
                    code: 8657,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.rightwards_double_arrow',
                    searchTags: '',
                    code: 8658,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.downwards_double_arrow',
                    searchTags: '',
                    code: 8659,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.left_right_double_arrow',
                    searchTags: '',
                    code: 8660,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.therefore',
                    searchTags: '',
                    code: 8756,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.subset_of',
                    searchTags: '',
                    code: 8834,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.superset_of',
                    searchTags: '',
                    code: 8835,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.not_a_subset_of',
                    searchTags: '',
                    code: 8836,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.subset_of_or_equal_to',
                    searchTags: '',
                    code: 8838,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.superset_of_or_equal_to',
                    searchTags: '',
                    code: 8839,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.circled_plus',
                    searchTags: '',
                    code: 8853,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.circled_times',
                    searchTags: '',
                    code: 8855,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.perpendicular',
                    searchTags: '',
                    code: 8869,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.dot_operator',
                    searchTags: '',
                    code: 8901,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.left_ceiling',
                    searchTags: '',
                    code: 8968,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.right_ceiling',
                    searchTags: '',
                    code: 8969,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.left_floor',
                    searchTags: '',
                    code: 8970,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.right_floor',
                    searchTags: '',
                    code: 8971,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.left_pointing_angle_bracket',
                    searchTags: '',
                    code: 9001,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.right_pointing_angle_bracket',
                    searchTags: '',
                    code: 9002,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.lozenge',
                    searchTags: '',
                    code: 9674,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.black_spade_suit',
                    searchTags: '',
                    code: 9824,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.black_club_suit',
                    searchTags: '',
                    code: 9827,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.black_heart_suit',
                    searchTags: '',
                    code: 9829,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.black_diamond_suit',
                    searchTags: '',
                    code: 9830,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.en_space',
                    searchTags: '',
                    code: 8194,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.em_space',
                    searchTags: '',
                    code: 8195,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.thin_space',
                    searchTags: '',
                    code: 8201,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.zero_width_non_joiner',
                    searchTags: '',
                    code: 8204,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.zero_width_joiner',
                    searchTags: '',
                    code: 8205,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.left_to_right_mark',
                    searchTags: '',
                    code: 8206,
                    duplicate: false,
                },
                {
                    label: 'enhavo_form.wysiwyg_form.command.special_characters.tabs.aarrows.right_to_left_mark',
                    searchTags: '',
                    code: 8207,
                    duplicate: false,
                },
            ],
        },
    ];

    public generateSearchTags(translator: Translator, form: WysiwygForm)
    {
        for(let tab of this.characters) {
            for(let character of tab.characters) {
                if (character.searchTags) {
                    character.searchTags = character.searchTags.toLowerCase();
                    continue;
                }
                character.searchTags = translator.trans(character.label, {}, this.getTranslationDomain(form)).toLowerCase();
            }
        }
    }
}

export class CharacterTab
{
    label: string;
    characters: Character[];
}

export class Character
{
    label: string;
    searchTags: string;
    code: number;
    duplicate: boolean;
}
