// @enhavo/app/vite/rollup-plugin-container-di.js
import { createFilter } from '@rollup/pluginutils';
import Compiler from '@enhavo/dependency-injection/compiler/Compiler.js';
import ContainerBuilder from '@enhavo/dependency-injection/container/ContainerBuilder.js';
import Loader from '@enhavo/dependency-injection/loader/Loader.js';
import fs from 'fs';

const defaults = {
    transform: null,
    extensions: ['.di.yaml']
};

export default function (opts = {}) {
    const options = Object.assign({}, defaults, opts);
    const { extensions } = options;
    const filter = createFilter(options.include, options.exclude);

    let builder = new ContainerBuilder;

    return {
        name: 'container-di',
        async transform(content, id) {
            if (!extensions.some((ext) => id.toLowerCase().endsWith(ext))) return null;
            if (!filter(id)) return null;

            let loader = new Loader();
            loader.loadFile(id, builder)

            for (let loadedFile of loader.loadedFiles) {
                if (fs.existsSync(loadedFile) && fs.lstatSync(loadedFile).isFile()) {
                    this.addWatchFile(loadedFile);
                }
            }

            await builder.prepare();
            let compiler = new Compiler;
            let resultData = compiler.compile(builder);

            return {
                code: resultData,
                map: null,
            };
        },
        outputOptions(outputOptions) {
            const chunks = {};
            for (let definition of builder.getDefinitions()) {
                let chunkName = definition.chunckName;
                if (chunkName) {
                    if (!chunks[chunkName]) {
                        chunks[chunkName] = [];
                    }
                    chunks[chunkName].push(definition.from ? definition.from : definition.name);
                }
            }

            outputOptions.manualChunks = chunks;

            outputOptions.manualChunks = {
                action: [
                    '@enhavo/app/action/ActionManager',
                    '@enhavo/app/action/model/CloseAction',
                    '@enhavo/app/action/model/DeleteAction',
                    '@enhavo/app/action/model/DownloadAction',
                    '@enhavo/app/action/model/DropdownAction',
                    '@enhavo/app/action/model/DuplicateAction',
                    '@enhavo/app/action/model/EventAction',
                    '@enhavo/app/action/model/FilterAction',
                    '@enhavo/app/action/model/ModalAction',
                    '@enhavo/app/action/model/FormAction',
                    '@enhavo/app/action/model/OpenAction',
                    '@enhavo/app/action/model/PreviewAction',
                    '@enhavo/app/action/model/SaveAction',
                    '@enhavo/app/action/model/AutoSaveAction',
                    '@enhavo/app/action/model/PreviewModeAction',
                    '@enhavo/app/action/model/RevisionAwareRestoreAction',
                    '@enhavo/app/action/model/ArchiveAction',
                    // '@enhavo/app/components/action/ActionAction.vue',
                    // '@enhavo/app/components/action/ActionPreview.vue',
                    // '@enhavo/app/components/action/ActionAutoSave.vue',
                    // '@enhavo/app/components/action/ActionDropdown.vue',
                    // '@enhavo/app/components/action/ActionFilter.vue',
                    '@enhavo/app/menu/MenuManager',
                    '@enhavo/media/action/MediaDownloadAction',
                    '@enhavo/media/action/MediaFormatAction',
                    '@enhavo/media-library/action/MediaLibraryUploadAction',
                    '@enhavo/media-library/action/MediaLibraryOpenSelectAction',
                    '@enhavo/media-library/action/MediaLibrarySelectAction'
                ],
                batch: [
                    '@enhavo/app/batch/BatchManager',
                    '@enhavo/app/batch/model/UrlBatch',
                    '@enhavo/app/batch/model/ModalBatch',
                    '@enhavo/app/batch/model/FormBatch',
                    '@enhavo/media-library/batch/MediaLibrarySelectBatch'
                ],
                // collection: [
                //     '@enhavo/app/collection/CollectionFactory',
                //     '@enhavo/app/collection/model/TableCollection',
                //     '@enhavo/app/collection/model/ListCollection',
                //     '@enhavo/app/components/collection/CollectionList.vue',
                //     '@enhavo/app/components/collection/CollectionListItem.vue',
                //     '@enhavo/app/components/collection/CollectionTable.vue',
                //     '@enhavo/app/components/collection/CollectionTableRow.vue',
                //     '@enhavo/app/components/collection/CollectionTablePagination.vue',
                //     '@enhavo/media-library/collection/MediaLibraryCollection'
                // ],
                // column: [
                //     '@enhavo/app/column/ColumnFactory',
                //     '@enhavo/app/column/ColumnManager',
                //     '@enhavo/app/column/model/BooleanColumn',
                //     '@enhavo/app/column/model/TextColumn',
                //     '@enhavo/app/column/model/ActionColumn',
                //     '@enhavo/app/column/model/SubColumn',
                //     '@enhavo/app/column/model/MediaColumn',
                //     '@enhavo/app/column/model/StateColumn',
                //     '@enhavo/dashboard/dashboard/DashboardWidgetFactory'
                // ],
                vue: [
                    '@enhavo/app/components/action/ActionBar.vue',
                    '@enhavo/app/components/flash-message/FlashMessages.vue',
                    '@enhavo/app/components/flash-message/FlashMessage.vue',
                    '@enhavo/app/components/modal/ModalStack.vue',
                    '@enhavo/app/components/modal/ModalForm.vue',
                    '@enhavo/app/components/modal/ModalIframe.vue',
                    '@enhavo/app/components/modal/ModalOutputStream.vue',
                    '@enhavo/app/components/filter/FilterBar.vue',
                    '@enhavo/app/components/ui/UiComponents.vue',
                    '@enhavo/app/components/main/OverlayContainer.vue',
                    '@enhavo/app/components/batch/BatchDropdown.vue',
                    '@enhavo/app/components/toolbar/ToolbarToolbar.vue',
                    '@enhavo/app/components/main/LoadingScreen.vue',
                    '@enhavo/app/components/main/MainMain.vue',
                    '@enhavo/app/components/app/AppApp.vue',
                    '@enhavo/app/components/resource/ResourceIndex.vue',
                    '@enhavo/app/components/resource/ResourceInput.vue',
                    '@enhavo/app/components/resource/ResourcePreview.vue',
                    // 'vue-select',
                    // '@vuepic/vue-datepicker',
                    '@enhavo/app/components/menu/MenuMenu.vue',
                    '@enhavo/vue-form/components/FormChoiceComponent.vue',
                    '@enhavo/vue-form/components/FormChoiceExpandedComponent.vue',
                    '@enhavo/vue-form/components/FormChoiceCollapsedComponent.vue',
                    '@enhavo/vue-form/components/FormChoiceOptionComponent.vue',
                    '@enhavo/vue-form/components/FormChoiceOptgroupComponent.vue',
                    '@enhavo/vue-form/components/FormCompoundWidgetComponent.vue',
                    '@enhavo/vue-form/components/FormErrorsComponent.vue',
                    '@enhavo/vue-form/components/FormRowComponent.vue',
                    '@enhavo/vue-form/components/FormRowsComponent.vue',
                    '@enhavo/vue-form/components/FormFormComponent.vue',
                    '@enhavo/vue-form/components/FormHelpComponent.vue',
                    '@enhavo/vue-form/components/FormSimpleComponent.vue',
                    '@enhavo/vue-form/components/FormLabelComponent.vue',
                    '@enhavo/vue-form/components/FormTextareaComponent.vue',
                    '@enhavo/vue-form/components/FormWidgetComponent.vue',
                    '@enhavo/vue-form/components/FormRadioComponent.vue',
                    '@enhavo/vue-form/components/FormCheckboxComponent.vue',
                    '@enhavo/vue-form/components/FormButtonComponent.vue',
                    '@enhavo/vue-form/components/FormSubmitComponent.vue',
                    '@enhavo/vue-form/components/FormButtonRowComponent.vue',
                    '@enhavo/vue-form/components/FormHiddenRowComponent.vue',
                    // '@enhavo/app/vue/VueRegistry',
                    // '@enhavo/app/vue/ClickOutside',
                    // '@enhavo/form/form/visitor/AdminFormVisitor',
                    '@enhavo/form/components/FormList.vue',
                    '@enhavo/form/components/FormDateTime.vue',
                    '@enhavo/form/components/FormListItem.vue',
                    '@enhavo/form/components/FormPolyCollection.vue',
                    '@enhavo/form/components/FormPolyCollectionItem.vue',
                    '@enhavo/form/components/FormWysiwyg.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygMenuGroup.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygMenuButton.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygMenuSubmenu.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygMenuSpacer.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygModalContainer.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygModalLink.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygModalSourceCode.vue',
                    '@enhavo/form/components/wysiwyg/FormWysiwygSearchReplaceInlineForm.vue',
                    '@enhavo/form/components/FormConditional.vue',
                    '@enhavo/form/components/FormSelect2ChoiceCollapsed.vue',
                    '@enhavo/form/components/FormAdminChoiceExpanded.vue',
                    '@enhavo/form/components/FormAdminChoice.vue',
                    '@enhavo/form/components/FormChoiceTree.vue',
                    '@enhavo/form/components/FormChoiceTreeRow.vue',
                    // '@enhavo/form/components/FormICheckCheckbox.vue',
                    // '@enhavo/form/components/FormICheckRadio.vue',
                    '@enhavo/form/components/FormAutoComplete.vue',
                    '@enhavo/form/components/FormHtmlTag.vue',
                    // '@enhavo/form/form/model/PolyCollectionForm',
                    // '@enhavo/form/form/model/ListForm',
                    // '@enhavo/form/form/model/WysiwygForm',
                    // '@enhavo/form/form/model/PositionForm',
                    // '@enhavo/form/form/model/ConditionalForm',
                    // '@enhavo/form/form/model/AutoCompleteForm',
                    // '@enhavo/form/form/model/DateTimeForm',
                    // '@enhavo/form/form/model/ChoiceTreeForm',
                    '@enhavo/block/components/tab/TabBlock.vue',
                    '@enhavo/block/components/form/FormBlockCollectionRow.vue',
                    '@enhavo/block/components/action/ActionBlockCollapse.vue',
                    '@enhavo/block/components/form/FormBlockNode.vue',
                    '@enhavo/media/components/FormMedia.vue',
                    '@enhavo/media/components/FormMediaItem.vue',
                    '@enhavo/media/components/ActionMediaForm.vue',
                    '@enhavo/media/components/ActionMediaFormat.vue',
                    // '@enhavo/media/resolver/FileResolver',
                    // '@enhavo/media/form/model/MediaForm',
                    // '@enhavo/media/form/model/MediaItemForm',
                    '@enhavo/media/components/MediaImageCropper.vue',
                    '@enhavo/media-library/components/collection/CollectionMediaLibrary.vue',
                    '@enhavo/media-library/components/collection/CollectionMediaLibraryThumbnail.vue',
                    '@enhavo/media-library/components/filter/FilterMediaLibrarySearch.vue',
                    '@enhavo/media-library/components/filter/FilterMediaLibraryChoice.vue',
                    '@enhavo/media-library/components/filter/FilterMediaLibraryCheckbox.vue'
                ],
                main: [
                    '@enhavo/app/components/toolbar/ToolbarWidgetQuickMenu.vue',
                    '@enhavo/app/components/toolbar/ToolbarWidgetIcon.vue',
                    '@enhavo/app/components/menu/MenuDropdown.vue',
                    '@enhavo/app/components/menu/MenuItem.vue',
                    '@enhavo/app/components/menu/MenuList.vue',
                    '@enhavo/app/components/menu/MenuNotification.vue',
                    '@enhavo/app/controllers/VueController',
                    '@enhavo/app/manager/MainManager',
                    '@enhavo/app/menu/model/BaseMenuItem',
                    '@enhavo/app/menu/model/ListMenuItem',
                    '@enhavo/app/menu/model/DropdownMenuItem',
                    '@enhavo/app/controllers/FormController',
                    '@enhavo/app/toolbar/ToolbarWidgetManager',
                    '@enhavo/app/toolbar/model/BaseToolbarWidget',
                    '@enhavo/app/toolbar/model/IconToolbarWidget',
                    '@enhavo/app/toolbar/model/NewWindowToolbarWidget',
                    '@enhavo/app/toolbar/model/QuickMenuToolbarWidget',
                    '@enhavo/multi-tenancy/menu/SwitchTenantMenuItem'
                ],
                frame: [
                    '@enhavo/app/components/frame/FrameStack.vue',
                    '@enhavo/app/components/frame/FrameStackFrame.vue',
                    '@enhavo/app/components/frame/FrameStackMobileDropdown.vue',
                    '@enhavo/app/frame/FrameStack',
                    '@enhavo/app/frame/FrameStackSubscriber',
                    '@enhavo/app/frame/FrameEventDispatcher',
                    '@enhavo/app/frame/FrameManager',
                    '@enhavo/app/frame/FrameArrangeManager',
                    '@enhavo/app/frame/FrameStateManager'
                ],
                grid: [
                    '@enhavo/app/components/column/ColumnBoolean.vue',
                    '@enhavo/app/components/column/ColumnText.vue',
                    '@enhavo/app/components/column/ColumnAction.vue',
                    '@enhavo/app/components/column/ColumnSub.vue',
                    '@enhavo/app/components/column/ColumnMedia.vue',
                    '@enhavo/app/components/column/ColumnState.vue',
                    '@enhavo/app/components/filter/FilterAutoComplete.vue',
                    '@enhavo/app/components/filter/FilterCheckbox.vue',
                    '@enhavo/app/components/filter/FilterDropdown.vue',
                    '@enhavo/app/components/filter/FilterText.vue',
                    '@enhavo/app/components/filter/FilterBetween.vue',
                    '@enhavo/app/components/filter/FilterDateBetween.vue',
                    '@enhavo/app/filter/FilterManager',
                    '@enhavo/app/filter/model/BooleanFilter',
                    '@enhavo/app/filter/model/TextFilter',
                    '@enhavo/app/filter/model/AutoCompleteEntityFilter',
                    '@enhavo/app/filter/model/EntityFilter',
                    '@enhavo/app/filter/model/OptionFilter',
                    '@enhavo/app/filter/model/BetweenFilter',
                    '@enhavo/app/filter/model/DateBetweenFilter'
                ],
                tab: [
                    '@enhavo/app/components/tab/TabForm.vue',
                    '@enhavo/app/components/tab/TabRevision.vue',
                    '@enhavo/app/components/tab/TabBase.vue',
                    '@enhavo/app/tab/TabManager',
                    '@enhavo/app/tab/model/BaseTab',
                    '@enhavo/app/tab/model/FormTab',
                    '@enhavo/app/tab/model/RevisionTab',
                    '@enhavo/block/tab/BlockTab',
                    '@enhavo/block/action/BlockCollapseAction'
                ],
                form: [
                    '@enhavo/app/components/form/FormAdminRow.vue',
                    '@enhavo/app/components/form/FormAdminWidget.vue',
                    '@enhavo/app/components/form/FormAdminHelp.vue',
                    '@enhavo/app/form/FormApp',
                    '@enhavo/app/form/FormRegistry',
                    '@enhavo/app/form/Form',
                    '@enhavo/app/form/visitor/AdminFormVisitor',
                    '@enhavo/vue-form/form/FormFactory',
                    '@enhavo/vue-form/form/FormEventDispatcher',
                    '@enhavo/vue-form/model/CheckboxForm',
                    '@enhavo/vue-form/model/RadioForm',
                    '@enhavo/vue-form/model/ChoiceForm',
                    // '@enhavo/form/loader/CheckboxLoader',
                    // '@enhavo/form/loader/SelectLoader',
                    // '@enhavo/form/loader/DateTimeLoader',
                    // '@enhavo/form/loader/DateLoader',
                    // '@enhavo/form/loader/ListLoader',
                    // '@enhavo/form/loader/AutoCompleteLoader',
                    // '@enhavo/form/loader/AutoSuggestLoader',
                    // '@enhavo/form/loader/WeekendDateLoader',
                    // '@enhavo/form/loader/PolyCollectionLoader',
                    // '@enhavo/form/loader/ConditionLoader',
                    '@enhavo/media-library/loader/MediaLibraryLoader',
                    '@enhavo/translation/form/loader/TranslationLoader'
                ],
                // delete: [
                //     '@enhavo/app/delete/DeleteApp',
                //     '@enhavo/app/components/delete/DeleteComponent.vue'
                // ],
                // 'expression-language': [ '@enhavo/app/expression-language/ExpressionLanguage' ],
                // 'flash-message': [ '@enhavo/app/flash-message/FlashMessenger' ],
                // modal: [
                //     '@enhavo/app/modal/ModalManager',
                //     '@enhavo/app/modal/model/IframeModal',
                //     '@enhavo/app/modal/model/FormModal',
                //     '@enhavo/app/modal/model/OutputStreamModal'
                // ],
                // app: [ '@enhavo/app/routing/Router' ],
                // stimulus: [ '@hotwired/stimulus' ],
                // view: [
                //     '@enhavo/app/translation/Translator',
                //     '@enhavo/core/Router',
                //     '@enhavo/core/Translator'
                // ],
                // controller: [
                //     '@enhavo/form/controllers/FormStyleController',
                //     '@enhavo/media/controllers/MediaStyleController'
                // ],
                // 'image-cropper': [
                //     '@enhavo/media/manager/ImageCropperManager',
                //     '@enhavo/media/action/MediaCropAction'
                // ],
                // 'media-library': [
                //     '@enhavo/media-library/components/media-library/MediaLibrary.vue',
                //     '@enhavo/media-library/manager/MediaLibraryManager'
                // ],
                // user: [
                //     '@enhavo/user/manager/UserManager',
                //     '@enhavo/user/components/form/FormUserGroupPermission.vue',
                //     '@enhavo/user/components/UserApp.vue',
                //     '@enhavo/user/components/UserLogin.vue',
                //     '@enhavo/user/components/reset-password/UserResetPasswordRequest.vue',
                //     '@enhavo/user/components/reset-password/UserResetPasswordConfirm.vue',
                //     '@enhavo/user/components/reset-password/UserResetPasswordCheck.vue',
                //     '@enhavo/user/components/reset-password/UserResetPasswordFinish.vue',
                //     '@enhavo/user/controllers/UserController'
                // ]
            }

            return outputOptions;
        }
    };
}
